splinterd/src/config/builder.rs

```diff
@@ -12,14 +12,441 @@
 }
 
+/// ConfigBuilder collects PartialConfig objects from various sources to be used to generate a
+/// Config object. The ConfigBuilder builds the Config object in a specific order determined by the
+/// ConfigSource of the PartialConfig.
+pub struct ConfigBuilder {
+    partial_configs: HashMap<ConfigSource, PartialConfig>,
+    storage: Option<(String, ConfigSource)>,
+    transport: Option<(String, ConfigSource)>,
+    cert_dir: Option<(String, ConfigSource)>,
+    ca_certs: Option<(String, ConfigSource)>,
+    client_cert: Option<(String, ConfigSource)>,
+    client_key: Option<(String, ConfigSource)>,
+    server_cert: Option<(String, ConfigSource)>,
+    server_key: Option<(String, ConfigSource)>,
+    service_endpoint: Option<(String, ConfigSource)>,
+    network_endpoint: Option<(String, ConfigSource)>,
+    peers: Option<(Vec<String>, ConfigSource)>,
+    node_id: Option<(String, ConfigSource)>,
+    bind: Option<(String, ConfigSource)>,
+    #[cfg(feature = "database")]
+    database: Option<(String, ConfigSource)>,
+    registry_backend: Option<(String, ConfigSource)>,
+    registry_file: Option<(String, ConfigSource)>,
+    heartbeat_interval: Option<(u64, ConfigSource)>,
+    admin_service_coordinator_timeout: Option<(Duration, ConfigSource)>,
+    state_dir: Option<(String, ConfigSource)>,
+}
+
+impl ConfigBuilder {
+    pub fn new() -> Self {
+        ConfigBuilder {
+            partial_configs: HashMap::new(),
+            storage: None,
+            transport: None,
+            cert_dir: None,
+            ca_certs: None,
+            client_cert: None,
+            client_key: None,
+            server_cert: None,
+            server_key: None,
+            service_endpoint: None,
+            network_endpoint: None,
+            peers: None,
+            node_id: None,
+            bind: None,
+            #[cfg(feature = "database")]
+            database: None,
+            registry_backend: None,
+            registry_file: None,
+            heartbeat_interval: None,
+            admin_service_coordinator_timeout: None,
+            state_dir: None,
+        }
+    }
+
+    /// Adds a PartialConfig to the ConfigBuilder object.
+    ///
+    /// # Arguments
+    ///
+    /// * `partial` - A PartialConfig object generated from any of the config modules.
+    ///
+    pub fn with_partial_config(&mut self, partial: PartialConfig) -> Result<(), ConfigError> {
+        // Validate the PartialConfig has a valid ConfigSource to associate with its values.
+        let source = match partial.source() {
+            Some(s) => Ok(s),
+            None => Err(ConfigError::UnsetSource(String::from(
+                "Source unset, unable to add to ConfigBuilder",
+            ))),
+        }?;
+        self.partial_configs.insert(source, partial);
+        Ok(())
+    }
+
+    /// Builds a Config object by incorporating the values from each PartialConfig object.
+    ///
+    pub fn build(mut self) -> Result<Config, ConfigError> {
```

Comment:
This should build the config based on the order the partialconfigs were added and should not explicitly consider the source itself.

Reply:
Are we not concerned about the source in the sense that config values from the command line should have precedence over the default values? This takes care of that concern, regardless of when the PartialConfigs are added.

Reply:
We are, but the code using the builder should determine the order. Adding another partial config shouldn't involve changing this implementation.